### PR TITLE
Bookmarks: Undo behavior

### PR DESCRIPTION
**Pull request by malmstein**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205171041444238/f

### Description
This PR changes how undo works at a UI layer level for saved sites.
Until now, when the user wanted to delete a bookmark, we were directly removing it from the DB. In the case that the user were to undo the change, we were adding a new row to the DB. This is problematic in many levels, changing order of Favourites and also triggering unnecessary Sync operations.

### Steps to test this PR
All test cases assume that Sync is enabled

_Remove Favourite from Bookmarks screen_
- [x] Add a favourite 
- [x] Verify that Sync operation is triggered
- [x] Open Bookmarks screen
- [x] Remove favourite
- [x] Verify Sync operation is not triggered
- [x] Verify Site does not appear in Favourites list
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site appears in the Favourites list
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Bookmark from Browser_
- [x] Add a bookmark 
- [x] Verify that Sync operation is triggered
- [x] Open Bookmarks screen
- [x] Remove bookmark
- [x] Verify Sync operation is not triggered
- [x] Verify Site does not appear in Favourites list
- [x] Verify Site does not appear in Bookmarks list
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site appears in the Favourites list
- [x] Verify Site appears in the Bookmarks list
- [x] Remove bookmark and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Favourite from Browser_
- [x] Add a favourite 
- [x] Verify that Sync operation is triggered
- [x] Remove favourite
- [x] Verify Sync operation is not triggered
- [x] Verify Site is not favourited in the Browser Menu
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site is favourited in the Browser Menu
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Bookmark from Browser_
- [x] Add a bookmark 
- [x] Verify that Sync operation is triggered
- [x] Open Browser Menu and tap on Edit
- [x] Remove bookmark
- [x] Verify Sync operation is not triggered
- [x] Verify Site is not favourited in the Browser Menu
- [x] Verify Site is not bookmarked in the Browser Menu (should see Add bookmark instead)
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site is favourited in the Browser Menu
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Favourite from New Tab_
- [x] Add a favourite 
- [x] Verify that Sync operation is triggered
- [x] Open New Tab
- [x] Long tap on Favourite and Delete it
- [x] Verify Sync operation is not triggered
- [x] Verify Site no longer appears in the grid
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site is visible in the grid
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Favourite from Omnibar Quick Access_
- [x] Add a favourite 
- [x] Verify that Sync operation is triggered
- [x] Tap on omnibar so that the quick access panel appears
- [x] Long tap on Favourite and Delete it
- [x] Verify Sync operation is not triggered
- [x] Verify Site no longer appears in the grid
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site is visible in the quick access panel
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

_Remove Favourite from Favourites Widget_
- [x] Add Favourites Widget to device Home screen
- [x] Add a favourite 
- [x] Verify that Sync operation is triggered
- [x] Send device to home and tap on Favourites Widget
- [x] Long tap on Favourite and Delete it
- [x] Verify Sync operation is not triggered
- [x] Verify Site no longer appears in the grid
- [x] Tap on Undo
- [x] Verify Sync operation is not triggered
- [x] Verify Site is visible in the quick access panel
- [x] Remove favourite and let Snackbar disappear
- [x] Verify that Sync operation is triggered

### CODE REVIEW

**Comment by malmstein** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3828** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3828" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3828?utm_source=stack-comment).

File: app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchActivity.kt [hunk 1]
```diff
@@ -391,15 +403,34 @@ class SystemSearchActivity : DuckDuckGoActivity() {
...
-        }.show()
+            viewModel.undoDelete(savedSite, oldPosition)
+        }
+            .addCallback(
+                object : BaseTransientBottomBar.BaseCallback<Snackbar>() {
+                    override fun onDismissed(
+                        transientBottomBar: Snackbar?,
+                        event: Int,
+                    ) {
+                        // when snackbar is not dismissed because of an action we want to
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchActivity.kt:424]: ditto. Just keep this agnostic of what should happen, and move logic to viewmodel.

File: app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchViewModel.kt [hunk 2]
```diff
@@ -328,25 +330,38 @@ class SystemSearchViewModel @Inject constructor(
...
-            }
-            else -> throw IllegalArgumentException("Illegal SavedSite to delete received")
+    private fun hideQuickAccessItem(quickAccessFavourite: FavoritesQuickAccessAdapter.QuickAccessFavorite): Int {
+        val quickAccessItems = resultsViewState.value as QuickAccessItems
+        val favourites = quickAccessItems.favorites.toMutableList()
+        val index = favourites.indexOf(quickAccessFavourite)
+        favourites.remove(quickAccessFavourite)
+
+        latestQuickAccessItems = QuickAccessItems(favourites)
+        resultsViewState.postValue(latestQuickAccessItems)
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchViewModel.kt:352]: same as commented in `BrowserTabViewmodel` should work. I would keep the observable approach, and avoid updating the state from here.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt [hunk 3]
```diff
@@ -142,19 +138,89 @@ class BookmarksViewModel @Inject constructor(
...
     }
 
-    fun insert(savedSite: SavedSite) {
-        viewModelScope.launch(dispatcherProvider.io()) {
-            savedSitesRepository.insert(savedSite)
+    fun undoDelete(savedSite: SavedSite) {
+        // we don't insert it, only flip the deleted flag
+        when (savedSite) {
+            is Bookmark -> {
+                val bookmarks = viewState.value?.bookmarks!!.toMutableList()
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt:190]: Posted a suggestion in `BrowserTabViewModel` to make this simpler emitting a list of hidden ids.

With that we don't need to flip deleted flag, and we all the logic will belong to the viewmodel.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt [hunk 4]
```diff
@@ -142,19 +138,89 @@ class BookmarksViewModel @Inject constructor(
...
     }
 
-    private fun delete(savedSite: SavedSite) {
+    private fun hide(savedSite: SavedSite) {
+        // we don't delete, just remove from the ui list
+        when (savedSite) {
+            is Bookmark -> {
+                val bookmarks = viewState.value?.bookmarks!!.toMutableList()
+                val index = bookmarks.indexOf(savedSite)
+                bookmarks[index] = savedSite.copy(deleted = "1")
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt:150]: ditto.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt [hunk 5]
```diff
@@ -175,10 +176,17 @@ class BookmarksActivity : DuckDuckGoActivity() {
         ) { viewState ->
             viewState?.let { state ->
                 if (parentId == SavedSitesNames.BOOKMARKS_ROOT) {
-                    favoritesAdapter?.setItems(state.favorites.map { FavoritesAdapter.FavoriteItem(it) })
+                    favoritesAdapter?.setItems(state.favorites.filter { it.deleted == null }.map { FavoritesAdapter.FavoriteItem(it) })
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt:179]: We should display whatever is in the ViewState. Any filter should be applied before emitting the from the viewmodel.

**Review by cmonfortep — COMMENTED** [review]
I have a few comments I would like to review together:
- Move logic out of the UI, and keep viewmodel methods agnostic.
- We don't have a consitent way to hide savedsites/folders. One implementation for Bookmarks screen, and different implementations for BrowserTab and SystemSearch screens. I think we should unify the logic (posted a proposal, let's review if that will work)

**Review by cmonfortep — APPROVED** [review]
LGTM works great!

**Pull request merged**